### PR TITLE
Impose new restrictions on paras init and cleanup

### PR DESCRIPTION
**Pull request by pepyakin**
First PR of the #3211 series.
Related https://github.com/paritytech/polkadot/pull/4322

For upcoming PVF pre-checking feature we will need to impose a couple of
new restrictions for:

- `schedule_para_initialize`.
- `schedule_para_cleanup`.

Specifically, for the former we do not want to allow registration of
wasm blob that is empty, i.e. 0 bytes. While that currently already
does not make a lot of sense, it allows us to simplify the PVF
pre-checking logic: if this PR is deployed before the following changes
for PVF prechecking then we can be sure that no paras onboarding have to
have to go through the PVF pre-checking. In case, we deploy it
altogether this property will allow us to distingush paras that came in
before PVF pre-checking.

For `schedule_para_cleanup` we do not want to allow offboarding of paras
that are undergoing the upgrade process. While this is not a harsh
restriction this change allows us to avoid making the PVF prechecking
more complicated than it has to be.

### CODE REVIEW

File: runtime/parachains/src/paras.rs [hunk 1]
```diff
@@ -1817,55 +1833,51 @@ mod tests {
...
-				assert_eq!(Paras::current_code(&para_id), Some(original_code.clone()));
-				check_code_is_stored(&original_code);
-				check_code_is_stored(&new_code);
+			// Enact the upgrade.
+			//
+			// For that run to block #7 and submit a new head.
+			assert_eq!(expected_at, 7);
+			run_to_block(7, None);
+			assert_eq!(<frame_system::Pallet<Test>>::block_number(), 7);
+			Paras::note_new_head(para_id, Default::default(), expected_at);
```
Comment (apopiak) [inline runtime/parachains/src/paras.rs:1845]:
```suggestion
			assert_ok!(Paras::note_new_head(para_id, Default::default(), expected_at));
```
assert ok?

File: runtime/common/src/paras_registrar.rs [hunk 2]
```diff
@@ -547,6 +549,7 @@ impl<T: Config> Pallet<T> {
 		parachain: bool,
 	) -> Result<(ParaGenesisArgs, BalanceOf<T>), sp_runtime::DispatchError> {
 		let config = configuration::Pallet::<T>::config();
+		ensure!(validation_code.0.len() > 0, Error::<T>::EmptyCode);
```
Comment (apopiak) [inline runtime/common/src/paras_registrar.rs:552]: Should we add a test?

**Review by apopiak — COMMENTED** [review]
looks fine to me, not sure I can judge the broader implications (because I'm not deeply familiar with the paras_registrar)

File: runtime/parachains/src/paras.rs (hunk 1 above)
Comment (pepyakin) [reply runtime/parachains/src/paras.rs]: ah, no, `note_new_head` does not even return a result, it's `Weight`

Comment (pepyakin) [reply runtime/parachains/src/paras.rs:1867]: I yanked that commit

File: runtime/parachains/src/paras.rs [hunk 3]
```diff
@@ -886,10 +888,24 @@ impl<T: Config> Pallet<T> {
 
 	/// Schedule a para to be cleaned up at the start of the next session.
 	///
-	/// Will return error if para is not a stable parachain or parathread.
+	/// Will return error if either is true:
+	///
+	/// - para is not a stable parachain or parathread.
```
Comment (bkchr) [inline runtime/parachains/src/paras.rs:893]: What is a stable parachain?

**Review by bkchr — APPROVED** [review]

File: runtime/parachains/src/paras.rs (hunk 3 above)
Comment (pepyakin) [reply runtime/parachains/src/paras.rs:893]: a stable parachain or parathread is a parachain which `ParaLifecycles::<T>::get(&para)` is either `Parachain` or `Parathread`, and not e.g. `Onboarding` or other transitory state.

Do you think this should be clarified?

Comment (bkchr) [reply runtime/parachains/src/paras.rs:893]: Ahh okay, never heard this before. 

`para is not an on-boarded parachain or parathread.`

Would maybe have been more clear to me. 

Or you could use something like
`[`ParaLifecycle::is_stable`] returns `false` for para`

This way we would directly link to the particular type and make it more clear?

Comment (pepyakin) [reply runtime/parachains/src/paras.rs:893]: Sure, good idea

**Comment by pepyakin** [comment]
Current dependencies on/for this PR:
* master
  * **PR #4371** 
    * **PR #4372** 
      * **PR #4360**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh) and will continue to be automatically updated while this PR remains open.

**Review by shawntabrizi — APPROVED** [review]
Looks good to me.

**Pull request merged**